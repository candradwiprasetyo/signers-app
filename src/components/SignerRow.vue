<template>
  <tr class="hover:bg-gray-50 transition">
    <td class="p-3 border-t">{{ formatNpwp(signer.npwp) }}</td>
    <td class="p-3 border-t">{{ signer.name }}</td>
    <td class="p-3 border-t">{{ signer.signatory === 'TAXPAYER' ? 'Wajib Pajak' : 'Kuasa' }}</td>
    <td class="p-3 border-t text-center">
      <StatusBadge :active="signer.statusTaxpayer === 'ACTIVE'" />
    </td>
    <td class="p-3 border-t">
      <DefaultBadge :isDefault="signer.defaultSignatory" />
    </td>
    <td class="p-3 border-t">
      <button
        @click="$emit('edit')"
        class="px-2 py-1 bg-white hover:bg-gray-200 border text-sm rounded mr-2 font-bold text-green-500"
      >
        <span class="material-icons text-sm"> edit </span>
      </button>
      <button
        @click="$emit('delete')"
        class="px-2 py-1 bg-white hover:bg-gray-200 border text-sm rounded mr-2 font-bold text-red-500"
      >
        <span class="material-icons text-sm"> delete </span>
      </button>
    </td>
  </tr>
</template>

<script setup lang="ts">
import StatusBadge from './StatusBadge.vue'
import DefaultBadge from './DefaultBadge.vue'

defineProps({
  signer: {
    type: Object,
    required: true,
  },
})

function formatNpwp(npwp: string) {
  if (!npwp) return '-'
  return npwp.replace(/^(\d{2})(\d{3})(\d{3})(\d{1})(\d{3})(\d{3})$/, '$1.$2.$3.$4-$5.$6')
}
</script>
