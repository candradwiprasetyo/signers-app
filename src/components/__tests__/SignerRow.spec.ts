import { describe, it, expect } from 'vitest'
import { mount } from '@vue/test-utils'
import SignerRow from '../SignerRow.vue'

const mockSigner = {
  id: '1',
  name: 'Candra Dwi',
  npwp: '756946380643000',
  signatory: 'TAXPAYER',
  statusTaxpayer: 'ACTIVE',
  defaultSignatory: true,
}

describe('SignerRow', () => {
  it('render correctly with given signer data', () => {
    const wrapper = mount(SignerRow, {
      props: { signer: mockSigner },
      global: {
        stubs: ['StatusBadge', 'DefaultBadge'],
      },
    })

    expect(wrapper.text()).toContain('75.694.638.0-643.000')
    expect(wrapper.text()).toContain('Candra Dwi')
    expect(wrapper.text()).toContain('Wajib Pajak')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('emit correctly when edit button is clicked', async () => {
    const wrapper = mount(SignerRow, {
      props: { signer: mockSigner },
      global: { stubs: ['StatusBadge', 'DefaultBadge'] },
    })

    const btn = wrapper.findAll('button')[0]
    await btn.trigger('click')
    expect(wrapper.emitted()).toHaveProperty('edit')
  })

  it('emit correctly when delete button is clicked', async () => {
    const wrapper = mount(SignerRow, {
      props: { signer: mockSigner },
      global: { stubs: ['StatusBadge', 'DefaultBadge'] },
    })

    const btn = wrapper.findAll('button')[1]
    await btn.trigger('click')
    expect(wrapper.emitted()).toHaveProperty('delete')
  })
})
