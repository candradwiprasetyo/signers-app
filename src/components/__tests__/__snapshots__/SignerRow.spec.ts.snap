// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SignerRow > render correctly with given signer data 1`] = `
"<tr class="hover:bg-gray-50 transition">
  <td class="p-3 border-t">75.694.638.0-643.000</td>
  <td class="p-3 border-t">Candra Dwi</td>
  <td class="p-3 border-t">Wajib Pajak</td>
  <td class="p-3 border-t text-center">
    <status-badge-stub active="true"></status-badge-stub>
  </td>
  <td class="p-3 border-t">
    <default-badge-stub isdefault="true"></default-badge-stub>
  </td>
  <td class="p-3 border-t"><button class="px-2 py-1 bg-white hover:bg-gray-200 border text-sm rounded mr-2 font-bold text-green-500"><span class="material-icons text-sm"> edit </span></button><button class="px-2 py-1 bg-white hover:bg-gray-200 border text-sm rounded mr-2 font-bold text-red-500"><span class="material-icons text-sm"> delete </span></button></td>
</tr>"
`;
